File: Programmers/Lv_1/대충만든자판.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

vector<int> solution(vector<string> keymap, vector<string> targets) {
    vector<int> answer;
    vector<int> v;

    for (int i = 0; i < targets.size(); i++) {
        int count = 0;
        for (int j = 0; j < targets[i].size(); j++) {
            bool found = false;
            for (int k = 0; k < keymap.size(); k++) {
                auto it = find(keymap[k].begin(), keymap[k].end(), targets[i][j]);
                if (it != keymap[k].end()) {
                    v.push_back(it - keymap[k].begin());
                    found = true;
                    continue;
                }
            }
            if (!found) {
                count = -1;
                break;
            }
            count += *min_element(v.begin(), v.end())+1;
            v.clear();
        }
        answer.push_back(count);
    }

    return answer;
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    int n;
    string data;
    vector<int> result;
    vector<string> keymap, targets;

    cin >> n;
    for (int i = 0; i < n; i++) {
        cin >> data;
        keymap.push_back(data);
    }

    for (int i = 0; i < n; i++) {
        cin >> data;
        targets.push_back(data);
    }

    result = solution(keymap, targets);
    for (int i = 0; i < result.size(); i++) {
        cout << result[i] << " ";
    }

    return 0;
}

File: Programmers/Lv_1/대충만든자판.js
function solution(keymap, targets) {
    const answer = [];
    const map = new Map();

    for (const key of keymap) {
        for (let i = 0; i < key.length; i++) {
            if (!map.has(key[i]) || i+1 < map.get(key[i])) map.set(key[i], i+1);
        }
    }

    for (const target of targets) {
        let count = 0;
        for (let i = 0; i < target.length; i++) {
            count += map.get(target[i]);
        }
        answer.push(count || -1);
    }

    return answer;
}

function solution2(keymap, targets) {
    const answer = [];
    const map = [];

    for (const items of keymap) {
        items.split('').map((item, index) => map[item] = (map[item] < index+1 ? map[item] : index+1));
    }

    for (const items of targets) {
        answer.push(items.split('').reduce((cur, item) => cur += mpa[item], 0) || -1);
    }

    return answer;
}